### PR TITLE
Introduce the prepare for checkpoint explicit gc mode

### PR DESCRIPTION
**Pull request by tajila**
Introduce J9MMCONSTANT_EXPLICIT_GC_PREPARE_FOR_CHECKPOINT explicit GC
mode.

Signed-off-by: Tobi Ajila <atobia@ca.ibm.com>

### CODE REVIEW

File: gc/base/MemorySubSpace.cpp [hunk 1]
```diff
@@ -936,8 +936,12 @@ MM_MemorySubSpace::systemGarbageCollect(MM_EnvironmentBase* env, uint32_t gcCode
...
+
+		if ((J9MMCONSTANT_EXPLICIT_GC_PREPARE_FOR_CHECKPOINT == gcCode)
 #if defined(OMR_GC_IDLE_HEAP_MANAGER)
-		if ((J9MMCONSTANT_EXPLICIT_GC_IDLE_GC == gcCode) && (_extensions->gcOnIdle)) {
+		|| ((J9MMCONSTANT_EXPLICIT_GC_IDLE_GC == gcCode) && (_extensions->gcOnIdle))
+#endif
+		) {
 			OMRPORT_ACCESS_FROM_ENVIRONMENT(env);
 			uint64_t startTime = omrtime_hires_clock();
 			uintptr_t releasedBytes = _extensions->heap->getDefaultMemorySpace()->releaseFreeMemoryPages(env);
```
Comment (amicic) [inline gc/base/MemorySubSpace.cpp:947]: If you want to use this, you'll have to change #ifdefs from a dozen of files that surround releaseFreeMemoryPages in various classes.
Either completely remove ifdefs or append them with a checkpoint specific build flag.

File: include_core/j9nongenerated.h [hunk 2]
```diff
@@ -135,6 +135,15 @@ typedef struct J9VMGC_SublistFragment {
...
 #define J9MMCONSTANT_IMPLICIT_GC_COMPLETE_CONCURRENT 13
+#define J9MMCONSTANT_EXPLICIT_GC_PREPARE_FOR_CHECKPOINT  14
+#if defined(OMR_GC_IDLE_HEAP_MANAGER)
+#define J9MM_IS_EXPLICIT_IDLE_GC(gcCode) ((gcCode) == J9MMCONSTANT_EXPLICIT_GC_IDLE_GC)
+#else
+#define J9MM_IS_EXPLICIT_IDLE_GC(gcCode) (1 == 0)
+#endif
+
+
+#define J9MM_IS_EXPLICIT_IDLE_OR_CHECKPOINT_GC(gcCode) (((gcCode) == J9MMCONSTANT_EXPLICIT_GC_PREPARE_FOR_CHECKPOINT) || J9MM_IS_EXPLICIT_IDLE_GC((gcCode)))
```
Comment (amicic) [inline include_core/j9nongenerated.h:146]: I don't see much value in this OR macro. Having a macro for individual J9MM_IS_EXPLICIT_IDLE_GC and even for IS_CHECKPOINT_GC, if they are frequently used, is probably ok.

File: gc/base/MemorySubSpace.cpp (hunk 1 above)
Comment (tajila) [reply gc/base/MemorySubSpace.cpp:947]: is `OMR_GC_IDLE_HEAP_MANAGER` on by default everywhere? I believe this is true on x86.

Comment (tajila) [reply gc/base/MemorySubSpace.cpp:947]: I could just remove that macro

**Comment by tajila** [comment]
@amicic changes are ready for another look

File: gc/base/MemorySubSpace.cpp (hunk 1 above)
Comment (amicic) [reply gc/base/MemorySubSpace.cpp:945]: I was thinking about removing ifdefs just from releaseFreeMemoryPages (and dependent APIs), so it's accessible for other higher level functionality like checkpoint. but still keeping them around anything that is very Idle GC specific (control variables like gcOnIdle, various constants, and any code that may use them).

**Review by amicic — APPROVED** [review]

**Comment by tajila** [comment]
@dsouzai Please review/merge these changes

**Review by dsouzai — APPROVED** [review]

**Comment by dsouzai** [comment]
jenkins build all

**Comment by dsouzai** [comment]
I probably shouldn't have run tests without looking at the azure pipeline first..

The code doesn't build as it stands:
```
/home/vsts/work/1/s/gc/base/MemoryPoolAddressOrderedListBase.cpp: In member function ‘uintptr_t MM_MemoryPoolAddressOrderedListBase::releaseFreeEntryMemoryPages(MM_EnvironmentBase*, MM_HeapLinkedFreeHeader*)’:
/home/vsts/work/1/s/gc/base/MemoryPoolAddressOrderedListBase.cpp:205:26: error: ‘class MM_GCExtensionsBase’ has no member named ‘idleMinimumFree’; did you mean ‘tlhMinimumSize’?
     if (0 < _extensions->idleMinimumFree) {
                          ^~~~~~~~~~~~~~~
                          tlhMinimumSize
/home/vsts/work/1/s/gc/base/MemoryPoolAddressOrderedListBase.cpp:206:60: error: ‘class MM_GCExtensionsBase’ has no member named ‘idleMinimumFree’; did you mean ‘tlhMinimumSize’?
      commitPagesCount = totalFreePagesCount * _extensions->idleMinimumFree / 100;
                                                            ^~~~~~~~~~~~~~~
                                                            tlhMinimumSize
make[2]: *** [gc/CMakeFiles/omrgc.dir/base/MemoryPoolAddressOrderedListBase.cpp.o] Error 1
make[1]: *** [gc/CMakeFiles/omrgc.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
gc/CMakeFiles/omrgc.dir/build.make:565: recipe for target 'gc/CMakeFiles/omrgc.dir/base/MemoryPoolAddressOrderedListBase.cpp.o' failed
```

**Comment by amicic** [comment]
So, let's make idleMinimumFree  more generic:
- rename it to decommitMinimumFree
- pull the declaration and construction out of #ifdef, so it exists always and defaults to 0 (which means decommit will be performed if there is any free memory)
- the easiest is: only IdleGC logic will be able to override the default value (leave the logic in GCExtensions.cpp wrapped around #ifdef), although we could go further and change the command line option more generic too

**Comment by tajila** [comment]
@amicic changes ready for another look

**Review by amicic — APPROVED** [review]

**Comment by amicic** [comment]
jenkins build all

**Comment by tajila** [comment]
z/OS failure

```
10:49:36  stderr: 
10:49:36  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2675)
10:49:36  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:2099)
10:49:36  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$500(CliGitAPIImpl.java:85)
10:49:36  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:619)
10:49:36  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:848)
10:49:36  	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:158)
10:49:36  	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:151)
10:49:36  	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
10:49:36  	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
10:49:36  	at hudson.remoting.Request$2.run(Request.java:376)
10:49:36  	at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:78)
10:49:36  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
10:49:36  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
10:49:36  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
10:49:36  	at java.base/java.lang.Thread.run(Thread.java:840)
10:49:36  	Suppressed: hudson.remoting.Channel$CallSiteStackTrace: Remote call to ZISVJD09
10:49:36  		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1784)
10:49:36  		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
10:49:36  		at hudson.remoting.Channel.call(Channel.java:1000)
10:49:36  		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:143)
10:49:36  		at jdk.internal.reflect.GeneratedMethodAccessor548.invoke(Unknown Source)
10:49:36  		at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
10:49:36  		at java.base/java.lang.reflect.Method.invoke(Method.java:566)
10:49:36  		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:129)
10:49:36  		at com.sun.proxy.$Proxy102.execute(Unknown Source)
10:49:36  		at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1226)
10:49:36  		at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1308)
10:49:36  		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:129)
10:49:36  		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:97)
10:49:36  		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:84)
10:49:36  		at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
10:49:36  		at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
10:49:36  		... 4 more
[Pipeline] }
```

**Comment by babsingh** [comment]
Relaunching should fix zOS failure.

**Comment by babsingh** [comment]
jenkins build zos

**Review by dsouzai — APPROVED** [review]

**Comment by dsouzai** [comment]
Merging given @amicic 's approval and the passing tests.

**Pull request merged**